lib/rules/no-undefined.js

```diff
@@ -21,15 +21,74 @@ module.exports = {
 
     create(context) {
 
+        /**
+         * Checks whether the given reference is 'undefined' or not.
+         *
+         * @param {escope.Reference} reference - The reference to check.
+         * @returns {boolean} `true` if the reference is 'undefined'.
+         */
+        function isUndefinedReference(reference) {
+            const id = reference.identifier;
+
+            return id && id.name === "undefined";
+        }
+
+        /**
+         * Report an invalid "undefined" identifier node.
+         * @param {ASTNode} node The node to report.
+         * @returns {void}
+         */
+        function report(node) {
+            context.report({
+                node,
+                message: "Unexpected use of undefined."
+            });
+        }
+
+        /**
+         * Checks the given scope for references to `undefined` and reports
+         * all references found.
+         * @param {escope.Scope} scope The scope to check.
+         * @returns {void}
+         */
+        function checkScope(scope) {
+            const undefinedVar = scope.set.get("undefined");
+
+            if (!undefinedVar) {
+                return;
+            }
+
+            const references = undefinedVar.references;
+
+            const defs = undefinedVar.defs;
+
+            // Report non-initializing references (those are covered in defs below)
+            references
+                .filter(ref => !ref.init)
+                .forEach(ref => report(ref.identifier));
+
+            defs.forEach(def => report(def.name));
+        }
+
         return {
+            "Program:exit"() {
+                const globalScope = context.getScope();
+
+                /* 'scope.through' includes all references to undefined
+                 * variables. If the variable 'undefined' is not defined, it
+                 * can be found in 'scope.through'.
+                 */
+                globalScope.through
+                    .filter(isUndefinedReference)
+                    .forEach(ref => report(ref.identifier));
+
+                const stack = [globalScope];
 
-            Identifier(node) {
-                if (node.name === "undefined") {
-                    const parent = context.getAncestors().pop();
+                while (stack.length) {
+                    const scope = stack.pop();
 
-                    if (!parent || parent.type !== "MemberExpression" || node !== parent.property || parent.computed) {
-                        context.report({ node, message: "Unexpected use of undefined." });
-                    }
+                    stack.push.apply(stack, scope.childScopes);
```

Comment:
I ran this command on my branch and on its merge-base:

```shell
TIMING=1 node ./bin/eslint.js --no-eslintrc --parser-options "ecmaVersion: 6" --rule "no-undefined: error" lib tests
```

On master, the rule ran for 280-330ms. On my branch, the rule ran for 44-48ms. I'm open to suggestions on how I can improve my testing in case I'm doing something wrong there, but I assume that the switch from "Identifier" visitor to "Program:exit" helps quite a bit.